File: examples/go/wslvault_example.go
// WSLVault Secret Lifecycle Example — Go
//
// Shows the full flow:
//   1. Write a secret (base64-encoded payload)
//   2. Read the secret back and decode it
//   3. List secrets
//   4. Read a specific version
//   5. Soft-delete a version
//   6. (Optional) Exchange an API key for a JWT via /v1/auth/api-key
//
// Usage:
//   VAULT_ADDR=http://localhost:8081 \
//   VAULT_TENANT_ID=019d813d-74bc-7660-89a7-f02fd9f2736d \
//   go run wslvault_example.go
//
// No external dependencies — uses only the standard library.

package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

const secretPath = "demo/go/service-token"

type writeRequest struct {
	Data string `json:"data"`
}

type writeResponse struct {
	SecretID string `json:"secret_id"`
	Version  int    `json:"version"`
}

type readResponse struct {
	Data      string `json:"data"`
	Version   int    `json:"version"`
	CreatedAt string `json:"created_at"`
}

type deleteRequest struct {
	Versions []int `json:"versions"`
}

// vaultClient holds the configuration for making requests to WSLVault.
type vaultClient struct {
	vaultAddr   string
	tenantID    string
	principalID string
	policies    string
	http        *http.Client
}

func newVaultClient(vaultAddr, tenantID, principalID, policies string) *vaultClient {
	return &vaultClient{
		vaultAddr:   vaultAddr,
		tenantID:    tenantID,
		principalID: principalID,
		policies:    policies,
		http:        &http.Client{},
	}
}

// do sends an authenticated HTTP request and decodes the JSON response into dest.
// Pass dest=nil when you don't need the response body (e.g. delete operations).
func (c *vaultClient) do(method, url string, body any, dest any) error {
	var reqBody io.Reader
	if body != nil {
		encoded, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("marshal request: %w", err)
		}
		reqBody = bytes.NewReader(encoded)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return fmt.Errorf("build request: %w", err)
	}

	// In production the gateway injects X-Principal-Id / X-Policies from the JWT.
	// When calling the secret-engine directly (dev / service mesh), pass them manually.
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Tenant-Id", c.tenantID)
	req.Header.Set("X-Principal-Id", c.principalID)
	req.Header.Set("X-Policies", c.policies)

	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("http %s %s: %w", method, url, err)
	}
	defer resp.Body.Close()

	respBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("http %s %s: status %d — %s", method, url, resp.StatusCode, respBytes)
	}

	if dest != nil && len(respBytes) > 0 {
		if err := json.Unmarshal(respBytes, dest); err != nil {
			return fmt.Errorf("decode response: %w", err)
		}
	}
	return nil
}

// Optional: exchange an API key for a JWT.
//
//	func exchangeAPIKey(identityAddr, apiKey, tenantID string) (string, error) {
//	    body := fmt.Sprintf(`{"api_key":"%s","tenant_id":"%s"}`, apiKey, tenantID)
//	    req, _ := http.NewRequest("POST", identityAddr+"/v1/auth/api-key",
//	        strings.NewReader(body))
//	    req.Header.Set("Content-Type", "application/json")
//	    resp, err := http.DefaultClient.Do(req)
//	    // ... parse resp.body["token"]
//	}

func main() {
	vaultAddr := envOrDefault("VAULT_ADDR", "http://localhost:8081")
	tenantID := envOrDefault("VAULT_TENANT_ID", "019d813d-74bc-7660-89a7-f02fd9f2736d")
	principalID := envOrDefault("VAULT_PRINCIPAL_ID", "go-example")
	policies := envOrDefault("VAULT_POLICIES", "admin")

	client := newVaultClient(vaultAddr, tenantID, principalID, policies)

	// ── 1. Write a secret ──────────────────────────────────────────────────────
	fmt.Printf("==> 1. Writing secret to '%s'...\n", secretPath)
	payload := map[string]string{
		"service_token": "svc-token-abc123",
		"environment":   "production",
	}
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		fatalf("marshal payload: %v", err)
	}
	data := base64.StdEncoding.EncodeToString(payloadJSON)

	var writeResp writeResponse
	if err := client.do("POST", vaultAddr+"/v1/secret/data/"+secretPath,
		writeRequest{Data: data}, &writeResp); err != nil {
		fatalf("write secret: %v", err)
	}
	fmt.Printf("    secret_id=%s  version=%d\n", writeResp.SecretID, writeResp.Version)

	// ── 2. Read the secret back ────────────────────────────────────────────────
	fmt.Printf("\n==> 2. Reading secret from '%s'...\n", secretPath)
	var readResp readResponse
	if err := client.do("GET", vaultAddr+"/v1/secret/data/"+secretPath, nil, &readResp); err != nil {
		fatalf("read secret: %v", err)
	}
	decoded, err := base64.StdEncoding.DecodeString(readResp.Data)
	if err != nil {
		fatalf("decode data: %v", err)
	}
	fmt.Printf("    Decoded: %s\n", decoded)

	// ── 3. List secrets ────────────────────────────────────────────────────────
	fmt.Println("\n==> 3. Listing secrets...")
	var listResp any
	if err := client.do("GET", vaultAddr+"/v1/secret/list", nil, &listResp); err != nil {
		fatalf("list secrets: %v", err)
	}
	listJSON, _ := json.MarshalIndent(listResp, "    ", "  ")
	fmt.Printf("    %s\n", listJSON)

	// ── 4. Read a specific version ─────────────────────────────────────────────
	fmt.Printf("\n==> 4. Reading version %d explicitly...\n", writeResp.Version)
	var verResp readResponse
	url := fmt.Sprintf("%s/v1/secret/data/%s?version=%d", vaultAddr, secretPath, writeResp.Version)
	if err := client.do("GET", url, nil, &verResp); err != nil {
		fatalf("read version: %v", err)
	}
	fmt.Printf("    version=%d  created_at=%s\n", verResp.Version, verResp.CreatedAt)

	// ── 5. Soft-delete the version ─────────────────────────────────────────────
	fmt.Printf("\n==> 5. Soft-deleting version %d...\n", writeResp.Version)
	if err := client.do("POST", vaultAddr+"/v1/secret/delete/"+secretPath,
		deleteRequest{Versions: []int{writeResp.Version}}, nil); err != nil {
		fatalf("delete secret: %v", err)
	}
	fmt.Println("    Deleted (soft — data retained for undelete).")

	fmt.Println("\nDone! All WSLVault operations completed successfully.")
}

func envOrDefault(key, defaultVal string) string {
	if v, ok := os.LookupEnv(key); ok && v != "" {
		return v
	}
	return defaultVal
}

func fatalf(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "ERROR: "+format+"\n", args...)
	os.Exit(1)
}

File: examples/nodejs/wslvault_example.js
#!/usr/bin/env node
// WSLVault Secret Lifecycle Example — Node.js (built-in fetch, Node 18+)
//
// Shows the full flow:
//   1. Write a secret (base64-encoded payload)
//   2. Read the secret back and decode it
//   3. List secrets
//   4. Read a specific version
//   5. Soft-delete a version
//   6. (Optional) Exchange an API key for a JWT via /v1/auth/api-key
//
// Usage:
//   VAULT_ADDR=http://localhost:8081 \
//   VAULT_IDENTITY_ADDR=http://localhost:18082 \
//   VAULT_TENANT_ID=019d813d-74bc-7660-89a7-f02fd9f2736d \
//   node wslvault_example.js

'use strict';

const VAULT_ADDR = process.env.VAULT_ADDR ?? 'http://localhost:8081';
const VAULT_IDENTITY_ADDR = process.env.VAULT_IDENTITY_ADDR ?? 'http://localhost:18082';
const VAULT_TENANT_ID = process.env.VAULT_TENANT_ID ?? '019d813d-74bc-7660-89a7-f02fd9f2736d';
const VAULT_PRINCIPAL_ID = process.env.VAULT_PRINCIPAL_ID ?? 'nodejs-example';
const VAULT_POLICIES = process.env.VAULT_POLICIES ?? 'admin';

const SECRET_PATH = 'demo/nodejs/api-key';

// In production the gateway injects X-Principal-Id / X-Policies from the JWT.
// When calling the secret-engine directly (dev / service mesh), pass them manually.
const vaultHeaders = {
  'Content-Type': 'application/json',
  'X-Tenant-Id': VAULT_TENANT_ID,
  'X-Principal-Id': VAULT_PRINCIPAL_ID,
  'X-Policies': VAULT_POLICIES,
};

// Optional: exchange an API key for a JWT
//
// async function authenticate(apiKey) {
//   const resp = await fetch(`${VAULT_IDENTITY_ADDR}/v1/auth/api-key`, {
//     method: 'POST',
//     headers: { 'Content-Type': 'application/json' },
//     body: JSON.stringify({ api_key: apiKey, tenant_id: VAULT_TENANT_ID }),
//   });
//   const { token } = await resp.json();
//   return token; // Use as: 'Authorization': `Bearer ${token}`
// }

/** Throw a descriptive error when the HTTP response signals failure. */
async function assertOk(resp, context) {
  if (!resp.ok) {
    const body = await resp.text();
    throw new Error(`${context}: HTTP ${resp.status} — ${body}`);
  }
  return resp;
}

async function main() {
  // ── 1. Write a secret ──────────────────────────────────────────────────────
  console.log(`==> 1. Writing secret to '${SECRET_PATH}'...`);
  const secretPayload = { api_key: 'sk-1234567890abcdef', service: 'openai', tier: 'premium' };
  // WSLVault stores opaque bytes — base64-encode the JSON payload.
  const data = Buffer.from(JSON.stringify(secretPayload)).toString('base64');

  const writeResp = await fetch(`${VAULT_ADDR}/v1/secret/data/${SECRET_PATH}`, {
    method: 'POST',
    headers: vaultHeaders,
    body: JSON.stringify({ data }),
  });
  await assertOk(writeResp, 'write secret');
  const writeBody = await writeResp.json();
  console.log(`    secret_id=${writeBody.secret_id}  version=${writeBody.version}`);
  const version = writeBody.version;

  // ── 2. Read the secret back ────────────────────────────────────────────────
  console.log(`\n==> 2. Reading secret from '${SECRET_PATH}'...`);
  const readResp = await fetch(`${VAULT_ADDR}/v1/secret/data/${SECRET_PATH}`, {
    headers: vaultHeaders,
  });
  await assertOk(readResp, 'read secret');
  const readBody = await readResp.json();
  const decoded = JSON.parse(Buffer.from(readBody.data, 'base64').toString('utf-8'));
  console.log('    Decoded:', decoded);

  // ── 3. List secrets ────────────────────────────────────────────────────────
  console.log('\n==> 3. Listing secrets...');
  const listResp = await fetch(`${VAULT_ADDR}/v1/secret/list`, { headers: vaultHeaders });
  await assertOk(listResp, 'list secrets');
  console.log('    Secrets:', await listResp.json());

  // ── 4. Read a specific version ─────────────────────────────────────────────
  console.log(`\n==> 4. Reading version ${version} explicitly...`);
  const verResp = await fetch(
    `${VAULT_ADDR}/v1/secret/data/${SECRET_PATH}?version=${version}`,
    { headers: vaultHeaders },
  );
  await assertOk(verResp, 'read version');
  const verBody = await verResp.json();
  console.log(`    version=${verBody.version}  created_at=${verBody.created_at}`);

  // ── 5. Soft-delete the version ─────────────────────────────────────────────
  console.log(`\n==> 5. Soft-deleting version ${version}...`);
  const delResp = await fetch(`${VAULT_ADDR}/v1/secret/delete/${SECRET_PATH}`, {
    method: 'POST',
    headers: vaultHeaders,
    body: JSON.stringify({ versions: [version] }),
  });
  await assertOk(delResp, 'delete secret');
  console.log('    Deleted (soft — data retained for undelete).');

  console.log('\nDone! All WSLVault operations completed successfully.');
}

main().catch((err) => {
  console.error('Error:', err.message);
  process.exit(1);
});
